pkg/apachepulsar/apachepulsar_test.go

```diff
@@ -0,0 +1,229 @@
+	}
+	if err := pool.Retry(func() error {
+		pulsarClient, err = pulsar.NewClient(pulsar.ClientOptions{
+			URL:               host,
+			OperationTimeout:  30 * time.Second,
+			ConnectionTimeout: 30 * time.Second,
+		})
+		if err != nil {
+			log.Fatalf("failed to create pulsar client: %v", err)
+		}
+		return nil
+	}); err != nil {
+		if resource != nil {
+			_ = pool.Purge(resource)
+		}
+		log.Fatalf("could not connect to apache pulsar %s", err)
+	}
+	defer pulsarClient.Close()
+	code := m.Run()
+	if resource != nil {
+		if err := pool.Purge(resource); err != nil {
+			log.Fatalf("Couln't purge resource %s", err)
+		}
+	}
+	os.Exit(code)
+}
+
+func TestPulsarSource_Read(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	log.Println("pulsar client", pulsarClient)
+	producer, err := initProducer(pulsarClient)
+	assert.Nil(t, err)
+	defer producer.Close()
+	go sendMessage(producer, ctx)
+	messageCh := make(chan sourcer.Message, 20)
+	consumer, err := pulsarClient.Subscribe(pulsar.ConsumerOptions{
+		Topic:            topic,
+		SubscriptionName: subscriptionName,
+		Type:             pulsar.Shared,
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer consumer.Close()
+	pulsarSource := NewPulsarSource(pulsarClient, consumer)
+	pulsarSource.Read(ctx, mocks.ReadRequest{
+		CountValue: 5,
+		Timeout:    20 * time.Second,
+	}, messageCh)
+	assert.Equal(t, 5, len(messageCh))
+	// Try reading 4 more messages
+	// Since the previous batch didn't get acked, the data source shouldn't allow us to read more messages
+	// We should get 0 messages, meaning the channel only holds the previous 5 messages
+
+	pulsarSource.Read(ctx, mocks.ReadRequest{
+		CountValue: 4,
+		Timeout:    time.Second,
+	}, messageCh)
+	assert.Equal(t, 5, len(messageCh))
+
+	msg1 := <-messageCh
+	msg2 := <-messageCh
+	msg3 := <-messageCh
+	msg4 := <-messageCh
+	msg5 := <-messageCh
+	fmt.Println(msg1.Offset().Value(), msg2.Offset().Value(), msg3.Offset().Value(), msg4.Offset().Value(), msg5.Offset().Value())
+
+	pulsarSource.Ack(ctx, mocks.TestAckRequest{
+		OffsetsValue: []sourcer.Offset{msg1.Offset(), msg2.Offset(), msg3.Offset(), msg4.Offset(), msg5.Offset()},
+	})
+
+	// Read 6 more messages
+	pulsarSource.Read(ctx, mocks.ReadRequest{
+		CountValue: 6,
+		Timeout:    time.Second,
+	}, messageCh)
+	assert.Equal(t, 6, len(messageCh))
+}
+
+func TestPulsarSource_Partitions(t *testing.T) {
+	os.Setenv("PULSAR_ADMIN_ENDPOINT", "http://localhost:8080")
+	os.Setenv("PULSAR_TOPIC", topic)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	consumer, err := pulsarClient.Subscribe(pulsar.ConsumerOptions{
+		Topic:            topic,
+		SubscriptionName: subscriptionName,
+		Type:             pulsar.Shared,
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer consumer.Close()
+	pulsarSource := NewPulsarSource(pulsarClient, consumer)
+	partitions := pulsarSource.Partitions(ctx)
+	assert.GreaterOrEqual(t, 1, len(partitions))
```

Comment:
with the current implementation, when will it be greater than 1? Can we verify the index value as well?